src/main/christophermanahan/github/io/carnitas/SocketInput.java

```diff
@@ -0,0 +1,34 @@
+package christophermanahan.github.io.carnitas;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.Socket;
+import java.util.Optional;
+
+public class SocketInput implements Input {
+
+  final Socket socket;
+
+  public SocketInput(Socket socket) {
+    this.socket = socket;
+  }
+
+  public Optional<String> receive() {
+    try {
+      return Optional.ofNullable(reader().readLine());
+    } catch (IOException e) {
+      throw new IllegalStateException("Could not read from socket input stream", e);
```

Comment:
We are returning an `Optional` here. Do we need to throw an exception? What would happen if we return `Optional.empty` instead? Could that work?

Reply:
@damonkelley The exception is thrown if the readLine operation failed. An empty optional would signify that the client has disconnected which would cause a readLine operation to fail. I believe returning the empty optional could be a way for the server to gracefully close upon a failed readLine?

Reply:
> An empty optional would signify that the connection has disconnected
I think it would more correct to say that an empty optional would signify that there was no data to read from the socket.